sympy/polys/matrices/domainmatrix.py

```diff
@@ -242,6 +241,18 @@ def from_Matrix(cls, M, **kwargs):
         Matrix
 
         """
+        m, n = M.shape
+        non_zero = 0
+
+        for i in range(m):
+            for j in range(n):
+                if not M[i, j].is_zero:
+                    non_zero += 1
+
+        if non_zero < (m*(n - 1) - 1)/2:
```

Comment:
Where does this criterion come from?

I think it would be better to choose a particular operation like say matrix multiplication or matrix addition or something and then time whether the dense or sparse implementations performs better given the domain in use.

Reply:
I say it on [wikipedia](https://en.wikipedia.org/wiki/Sparse_matrix), the CSR format uses this. Although there isn't any hard and fast rule for classifying a matrix as sparse.

Reply:
I think we should decide based on how sparse a matrix needs to be in order to make certain operations faster. So there should be some timings for basic matrix operations like `A + B` over `ZZ` and also over some more expensive domain.